File: Schema/SchemaDeploy/tables/Group/CGroupList.customisation.cs
using System;
using System.Text;
using System.Data;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Web;

using Framework;

namespace SchemaDeploy
{
	//Collection Class (Customisable half)
	public partial class CGroupList
	{
		#region Filters
		public CGroup GetOrCreate(string name)
		{
			if (string.IsNullOrEmpty(name))
				return null;

			var g = GetByName(name);
			if (null != g)
				return g;

			g = new CGroup();
			g.GroupName = name;
			g.Save();
			return g;
		}
		public CGroup GetByName(string name)
		{
			if (null == name)
				return null;
			name = name.ToLower();
			foreach (var i in this)
				if (i.GroupName.ToLower() == name)
					return i;
			return null;
		}
		#endregion

		#region Aggregation
		public CKeyList Keys
		{
			get
			{
				var temp = new CKeyList();
				foreach (var i in this)
					temp.AddRange(i.Keys);
				return temp;
			}
		}
		#endregion

		#region Searching (Optional)
		//Represents a simple search box to search PK and any string columns (add overloads as required, based on the pattern below)
		//e.g. public CGroupList Search(string nameOrId, ) { ...
		public CGroupList Search(string nameOrId)
		{
			//1. Normalisation
			nameOrId = (nameOrId ?? string.Empty).Trim().ToLower();

			//2. Start with a complete list
			CGroupList results = this;

			//3. Use any available index, such as those generated for fk/bool columns
			//Normal Case - non-unique index (e.g. foreign key)
			//if (int.MinValue != foreignKey) results = results.GetByForeignKey(foreignKey);

			//Special case - unique index (e.g. primary key)
			/*
            if (!string.IsNullOrEmpty(nameOrId)) 
            {
                int id;
                if (int.TryParse(nameOrId, out id))
                {
                    CGroup obj = this.GetById(id);
                    if (null != obj)
                    {
                        results = new CGroupList(1);
                        results.Add(obj);
                        return results;
                    }
                }
            }
            */

			//4. Exit early if remaining (non-index) filters are blank
			if (string.IsNullOrEmpty(nameOrId)) return results;

			//5. Manually search each record using custom match logic, building a shortlist
			CGroupList shortList = new CGroupList();
			foreach (CGroup i in results)
				if (Match(nameOrId, i))
					shortList.Add(i);
			return shortList;
		}
		//Manual Searching e.g for string-based columns i.e. anything not indexed (add more params if required)
		private bool Match(string name, CGroup obj)
		{
			if (!string.IsNullOrEmpty(name)) //Match any string column
			{
				if (null != obj.GroupName && obj.GroupName.ToLower().Contains(name)) return true;
				return false;   //If filter is active, reject any items that dont match
			}
			return true;    //No active filters (should catch this in step #4)
		}
		#endregion

		#region Cloning
		public CGroupList Clone(CDataSrc target) //, int parentId)
		{
			//No Transaction
			if (target is CDataSrcRemote)
				return Clone(target, null); //, parentId);

			//Transaction
			using (IDbConnection cn = target.Local.Connection())
			{
				IDbTransaction tx = cn.BeginTransaction();
				try
				{
					CGroupList clone = Clone(target, tx); //, parentId);
					tx.Commit();
					return clone;
				}
				catch
				{
					tx.Rollback();
					throw;
				}
			}
		}
		public CGroupList Clone(CDataSrc target, IDbTransaction txOrNull) //, int parentId)
		{
			CGroupList list = new CGroupList(this.Count);
			foreach (CGroup i in this)
				list.Add(i.Clone(target, txOrNull)); //, parentId));  *Child entities must reference the new parent
			return list;
		}
		#endregion

		#region Export to Csv
		//Web - Need to add a project reference to System.Web, or comment out these two methods
		public void ExportToCsv(HttpResponse response) { ExportToCsv(response, "Groups.csv"); }
		public void ExportToCsv(HttpResponse response, string fileName)
		{
			CDataSrc.ExportToCsv(response, fileName); //Standard response headers
			StreamWriter sw = new StreamWriter(response.OutputStream);
			ExportToCsv(sw);
			sw.Flush();
			response.End();
		}

		//Non-web
		public void ExportToCsv(string filePath)
		{
			StreamWriter sw = new StreamWriter(filePath);
			ExportToCsv(sw);
			sw.Close();
		}

		//Logic
		protected void ExportToCsv(StreamWriter sw)
		{
			string[] headings = new string[] { "GroupId", "GroupName" };
			CDataSrc.ExportToCsv(headings, sw);
			foreach (CGroup i in this)
			{
				object[] data = new object[] { i.GroupId, i.GroupName };
				CDataSrc.ExportToCsv(data, sw);
			}
		}
		#endregion
	}
}

File: Schema/SchemaConfig/tables/Group/CGroupList.customisation.vb
Imports System
Imports System.Data
Imports System.Collections.Generic

Imports Framework

'Collection Class (Customisable half)
Partial Public Class CGroupList

#Region "Alternative PK (on GroupName)"
    Default Public Overloads ReadOnly Property Item(ByVal name As String) As CGroup
        Get
            Return GetByName(name)
        End Get
    End Property
    Public Function GetByName(ByVal name As String) As CGroup
        Dim c As CGroup = Nothing
        IndexByName.TryGetValue(Normalise(name), c)
        Return c
    End Function
    <NonSerialized()> _
    Private _indexByName As Dictionary(Of String, CGroup)
    Private Function Normalise(ByVal key As String) As String
        Return LCase(Trim(key))
    End Function
    Private ReadOnly Property IndexByName() As Dictionary(Of String, CGroup)
        Get
            If Not IsNothing(_indexByName) Then
                If _indexByName.Count = Me.Count Then
                    Return _indexByName
                End If
            End If
            _indexByName = New Dictionary(Of String, CGroup)(Me.Count)
            For Each i As CGroup In Me
                _indexByName(Normalise(i.GroupName)) = i
            Next
            Return _indexByName
        End Get
    End Property
#End Region

#Region "Search"
    'Custom (build filters using parameters supplied, based around common combinations)
    'Public Function Search(ByVal name As String) As CGroupList
    '    Return Search(New CGroup.CSearchFilters(name))
    'End Function

    'Standard
    Public Function Search(ByVal filters As CGroup.CSearchFilters) As CGroupList
        Dim results As CGroupList = Me 'Start with a complete list
        
        With filters
            'Use any available index for fk/bool (e.g results=results.GetByFk)

            'Exit early if remaining filters are n/a
            If .NonIndexedFiltersAreNull Then Return results
        End With
            
        'Use match method for string searching or complex expressions
        Dim shortList As New CGroupList()
        For Each i As CGroup In results
            If i.Match(filters) Then shortList.Add(i)
        Next
        Return shortList
    End Function
#End Region

End Class
